client/python/cache_map.py

```diff
@@ -0,0 +1,176 @@
+    def __eq__(self, other):
+        if (isinstance(other, _InternalCacheObject)):
+            return self.__dict__ == other.__dict__
+        return False
+
+    def __hash__(self):
+        return hash(repr(self))
+
+    def __repr__(self):
+        return type(self).__name__ + str(self.__dict__)
+
+"""
+object returned by get method; type of cache_map values
+"""
+class CacheResult(_InternalCacheObject):
+    def __init__(self, value_time, cached_time, value):
+        self.value_time = value_time
+        self.cached_time = cached_time
+        self.value = value
+
+"""
+dictionary of cache entries implementing LRU cache
+"""
+class CacheMap(object):
+
+    class _CacheKey(_InternalCacheObject):
+        def __init__(self, table_id, key_id):
+            self.table_id = table_id
+            self.key_id = key_id
+
+    class _CacheEntry(_InternalCacheObject):
+        def __init__(self, cache_key, cache_result):
+            self.cache_key = cache_key
+            self.cache_result = cache_result
+
+    def __init__(self, max_size):
+        self.max_size = max_size
+
+        # Store values as a dictionary of lists:
+        # { (table_id, key_id): 
+        #       [ (value_time, cached_time, value)]}
+        self.cache_map = {}
+
+        # Track the order of entries added to / accessed in 
+        # cache to maintain LRU policy
+        # [ (table_id, key_id) ]
+        # TODO Remove self.lru_list = []
+        self.lru_list = dllist()
+
+    def _lru_add(self, entry):
+        self.lru_list.appendleft(entry)
+        self._lru_evict()
+
+    def _lru_refresh(self, arg):
+        if (type(arg) == int):
+            i = arg
+            lru_list = self.lru_list
+            lru_list.appendleft(lru_list.remove(lru_list.nodeat(i)))
+            self.lru_list = lru_list
+        elif (type(arg) == self._CacheEntry):
+            entry = arg
+            lru_list = self.lru_list
+            for i in xrange(len(lru_list)):
+                current_entry = lru_list.nodeat(i).value
+                if (current_entry == entry):
+                    self._lru_refresh(i)
+                    break
+
+    def _lru_evict(self):
+        cache_map = self.cache_map
+        lru_list = self.lru_list
+        max_size = self.max_size
+        remove_index = len(lru_list) - max(0, len(lru_list)-max_size)
+        for i in xrange(remove_index, len(lru_list)):
+            # Remove evicted entries from the cache_map
+            node = lru_list.nodeat(i)
+            entry = node.value
+            cache_key = entry.cache_key
+            cache_result = entry.cache_result
+            if (cache_key in cache_map):
+                cache_values = cache_map[cache_key]
+                cache_values.remove(cache_result)
+            # Reset the lru_list
+            lru_list.remove(node)
+        self.lru_list = lru_list
+
+    def _lru_remove(self, entries): 
+        lru_list = self.lru_list
+        for node in lru_list:
```

Comment:
This looks like it's O(N). Removing from a doubly-linked list should be O(1). LRU eviction should be O(1).